File: app/src/main/kotlin/tech/mmarca/openvitals/domain/model/ActivityBackfill.kt
package tech.mmarca.openvitals.domain.model

import kotlin.math.roundToLong
import tech.mmarca.openvitals.core.geo.haversineMeters
import tech.mmarca.openvitals.domain.insights.RouteElevation

internal fun ExerciseData.withRouteBackfilledMetrics(): ExerciseData {
    val metrics = route.takeIf { it.status == ExerciseRouteStatus.DATA }
        ?.points
        .orEmpty()
        .routeBackfillMetrics()

    return copy(
        totalDistanceMeters = totalDistanceMeters.backfilledBy(
            metrics.distanceMeters.takeIf { it >= MinBackfillDistanceMeters },
        ),
        elevationGainedMeters = elevationGainedMeters.backfilledBy(
            metrics.elevationGainMeters.takeIf { metrics.hasAltitudeData && it >= MinBackfillElevationMeters },
        ),
    )
}

internal fun ExerciseData.withSampleBackfilledMetrics(
    heartRateSamples: List<HeartRateSample>,
    speedSamples: List<SpeedSample>,
    cadenceSamples: List<ActivityCadenceSample>,
): ExerciseData =
    copy(
        // A watch that records speed but no DistanceRecord still implies a distance.
        totalDistanceMeters = totalDistanceMeters.backfilledBy(
            distanceFromSpeedSamples(speedSamples),
        ),
        averageHeartRateBpm = averageHeartRateBpm.backfilledBy(
            heartRateSamples
                .mapNotNull { it.beatsPerMinute.takeIf { bpm -> bpm > 0L } }
                .averageLongOrNull()
                ?.roundToLong(),
        ),
        averageSpeedMetersPerSecond = averageSpeedMetersPerSecond.backfilledBy(
            speedSamples
                .mapNotNull { it.metersPerSecond.takeIf { speed -> speed > 0.0 && speed.isFinite() } }
                .averageDoubleOrNull(),
        ),
        averageStepsCadenceRate = averageStepsCadenceRate.backfilledBy(
            cadenceSamples
                .filter { it.kind == ActivityCadenceKind.STEPS }
                .mapNotNull { it.rate.takeIf { rate -> rate > 0.0 && rate.isFinite() } }
                .averageDoubleOrNull(),
        ),
        averageCyclingCadenceRpm = averageCyclingCadenceRpm.backfilledBy(
            cadenceSamples
                .filter { it.kind == ActivityCadenceKind.CYCLING }
                .mapNotNull { it.rate.takeIf { rate -> rate > 0.0 && rate.isFinite() } }
                .averageDoubleOrNull(),
        ),
    )

private data class RouteBackfillMetrics(
    val distanceMeters: Double = 0.0,
    val elevationGainMeters: Double = 0.0,
    val altitudePairCount: Int = 0,
) {
    val hasAltitudeData: Boolean get() = altitudePairCount > 0
}

private fun List<ExerciseRoutePoint>.routeBackfillMetrics(): RouteBackfillMetrics {
    if (size < 2) return RouteBackfillMetrics()

    var distanceMeters = 0.0
    var altitudePairCount = 0

    val ordered = sortedBy { it.time }
    ordered.zipWithNext().forEach { (start, end) ->
        distanceMeters += start.distanceMetersTo(end)

        if (start.altitudeMeters != null && end.altitudeMeters != null) {
            altitudePairCount += 1
        }
    }

    // Through the smoothing filter, as a live recording is; raw rises bank GPS noise.
    val elevationGainMeters = RouteElevation.routeElevationGain(ordered)

    return RouteBackfillMetrics(
        distanceMeters = distanceMeters.takeIf { it.isFinite() } ?: 0.0,
        elevationGainMeters = elevationGainMeters.takeIf { it.isFinite() } ?: 0.0,
        altitudePairCount = altitudePairCount,
    )
}

private fun ExerciseRoutePoint.distanceMetersTo(other: ExerciseRoutePoint): Double =
    haversineMeters(latitude, longitude, other.latitude, other.longitude)

/** The distance a run of speed samples implies, by trapezoidal integration. Null when none. */
internal fun distanceFromSpeedSamples(samples: List<SpeedSample>): Double? {
    if (samples.size < 2) return null
    var cumulative = 0.0
    samples.sortedBy { it.time }.zipWithNext().forEach { (previous, current) ->
        val seconds = java.time.Duration.between(previous.time, current.time).toNanos() / 1_000_000_000.0
        val v0 = previous.metersPerSecond.coerceAtLeast(0.0)
        val v1 = current.metersPerSecond.coerceAtLeast(0.0)
        if (seconds > 0 && v0.isFinite() && v1.isFinite()) {
            cumulative += (v0 + v1) / 2.0 * seconds
        }
    }
    return cumulative.takeIf { it.isFinite() && it > 0.0 }
}

private fun Double?.backfilledBy(value: Double?): Double? =
    if (isMissingMetric() && value != null && value > 0.0 && value.isFinite()) value else this

private fun Long?.backfilledBy(value: Long?): Long? =
    if ((this == null || this <= 0L) && value != null && value > 0L) value else this

private fun Double?.isMissingMetric(): Boolean =
    this == null || this <= 0.0 || !isFinite()

private fun List<Long>.averageLongOrNull(): Double? =
    takeIf { it.isNotEmpty() }?.map { it.toDouble() }?.average()

private fun List<Double>.averageDoubleOrNull(): Double? =
    takeIf { it.isNotEmpty() }?.average()

private const val MinBackfillDistanceMeters = 1.0
private const val MinBackfillElevationMeters = 1.0

File: lib/domain/model/activity_backfill.dart
import '../../core/geo/geo_distance.dart';
import '../insights/activity_splits.dart';
import 'activity_models.dart';
import 'exercise_session_metrics.dart';
import 'heart_models.dart';
import '../../core/stats/stats.dart';

const double _minBackfillDistanceMeters = 1.0;
const double _minBackfillElevationMeters = 1.0;
const double _minBackfillElevationDeltaMeters = 1.0;

extension ActivityBackfill on ExerciseData {
  ExerciseData withRouteBackfilledMetrics() {
    final points = route.status == ExerciseRouteStatus.data
        ? route.points
        : const <ExerciseRoutePoint>[];
    final metrics = _routeBackfillMetrics(points);

    return copyWith(
      totalDistanceMeters: _backfilledByDouble(
        totalDistanceMeters,
        metrics.distanceMeters >= _minBackfillDistanceMeters
            ? metrics.distanceMeters
            : null,
      ),
      elevationGainedMeters: _backfilledByDouble(
        elevationGainedMeters,
        (metrics.hasAltitudeData &&
                metrics.elevationGainMeters >= _minBackfillElevationMeters)
            ? metrics.elevationGainMeters
            : null,
      ),
    );
  }

  /// Fill in the totals a session record does not carry, read from the sibling
  /// records covering its window (see [ExerciseSessionMetrics]).
  ///
  /// Health Connect is the authority here, so this runs BEFORE the sample-derived
  /// backfill: a real DistanceRecord beats a distance integrated from speed. Only
  /// missing fields are touched -- a figure the session already stated is never
  /// overwritten.
  ExerciseData withSessionMetricsBackfilled(ExerciseSessionMetrics metrics) =>
      copyWith(
        totalDistanceMeters: _backfilledByDouble(
          totalDistanceMeters,
          metrics.totalDistanceMeters,
        ),
        averageSpeedMetersPerSecond: _backfilledByDouble(
          averageSpeedMetersPerSecond,
          metrics.averageSpeedMetersPerSecond,
        ),
        steps: _backfilledByInt(steps, metrics.steps),
        totalCaloriesKcal: _backfilledByDouble(
          totalCaloriesKcal,
          metrics.totalCaloriesKcal,
        ),
        activeCaloriesKcal: _backfilledByDouble(
          activeCaloriesKcal,
          metrics.activeCaloriesKcal,
        ),
        elevationGainedMeters: _backfilledByDouble(
          elevationGainedMeters,
          metrics.elevationGainedMeters,
        ),
        floorsClimbed: _backfilledByInt(floorsClimbed, metrics.floorsClimbed),
        wheelchairPushes: _backfilledByInt(
          wheelchairPushes,
          metrics.wheelchairPushes,
        ),
      );

  ExerciseData withSampleBackfilledMetrics({
    required List<HeartRateSample> heartRateSamples,
    required List<SpeedSample> speedSamples,
    required List<ActivityCadenceSample> cadenceSamples,
  }) {
    final heartRateAverage = average(
      heartRateSamples
          .map((sample) => sample.beatsPerMinute)
          .where((bpm) => bpm > 0)
          .toList(),
    );
    return copyWith(
      // A watch that records speed but writes no DistanceRecord leaves the session
      // with no distance at all -- while the splits card, integrating these very
      // samples, cheerfully reports "every 1 km". Integrate them here too, so the
      // two can never disagree.
      totalDistanceMeters: _backfilledByDouble(
        totalDistanceMeters,
        distanceFromSpeedSamples(speedSamples),
      ),
      averageHeartRateBpm: _backfilledByInt(
        averageHeartRateBpm,
        heartRateAverage?.round(),
      ),
      averageSpeedMetersPerSecond: _backfilledByDouble(
        averageSpeedMetersPerSecond,
        average(
          speedSamples
              .map((sample) => sample.metersPerSecond)
              .where((speed) => speed > 0.0 && speed.isFinite)
              .toList(),
        ),
      ),
      averageStepsCadenceRate: _backfilledByDouble(
        averageStepsCadenceRate,
        average(
          cadenceSamples
              .where((sample) => sample.kind == ActivityCadenceKind.steps)
              .map((sample) => sample.rate)
              .where((rate) => rate > 0.0 && rate.isFinite)
              .toList(),
        ),
      ),
      averageCyclingCadenceRpm: _backfilledByDouble(
        averageCyclingCadenceRpm,
        average(
          cadenceSamples
              .where((sample) => sample.kind == ActivityCadenceKind.cycling)
              .map((sample) => sample.rate)
              .where((rate) => rate > 0.0 && rate.isFinite)
              .toList(),
        ),
      ),
    );
  }
}

class _RouteBackfillMetrics {
  const _RouteBackfillMetrics({
    this.distanceMeters = 0.0,
    this.elevationGainMeters = 0.0,
    this.altitudePairCount = 0,
  });

  final double distanceMeters;
  final double elevationGainMeters;
  final int altitudePairCount;

  bool get hasAltitudeData => altitudePairCount > 0;
}

_RouteBackfillMetrics _routeBackfillMetrics(List<ExerciseRoutePoint> points) {
  if (points.length < 2) return const _RouteBackfillMetrics();

  var distanceMeters = 0.0;
  var elevationGainMeters = 0.0;
  var altitudePairCount = 0;

  final sorted = [...points]..sort((a, b) => a.time.compareTo(b.time));
  for (var index = 0; index < sorted.length - 1; index++) {
    final start = sorted[index];
    final end = sorted[index + 1];
    distanceMeters += _distanceMetersTo(start, end);

    final startAltitude = start.altitudeMeters;
    final endAltitude = end.altitudeMeters;
    if (startAltitude != null && endAltitude != null) {
      altitudePairCount += 1;
      final gain = endAltitude - startAltitude;
      if (gain >= _minBackfillElevationDeltaMeters) {
        elevationGainMeters += gain;
      }
    }
  }

  return _RouteBackfillMetrics(
    distanceMeters: distanceMeters.isFinite ? distanceMeters : 0.0,
    elevationGainMeters: elevationGainMeters.isFinite ? elevationGainMeters : 0.0,
    altitudePairCount: altitudePairCount,
  );
}

double _distanceMetersTo(ExerciseRoutePoint point, ExerciseRoutePoint other) =>
    haversineMeters(
      point.latitude,
      point.longitude,
      other.latitude,
      other.longitude,
    );

double? _backfilledByDouble(double? current, double? value) {
  final isMissing = current == null || current <= 0.0 || !current.isFinite;
  if (isMissing && value != null && value > 0.0 && value.isFinite) {
    return value;
  }
  return current;
}

int? _backfilledByInt(int? current, int? value) {
  if ((current == null || current <= 0) && value != null && value > 0) {
    return value;
  }
  return current;
}

